### DISCUSSION TITLE
Slow download on a 10G ethernet connection

### DISCUSSION THREAD

**Discussion by upsj** [ipxe/ipxe#1288] [Q&A] [answered]
I tried building iPXE from source to replace the iPXE 1.0.0+ binaries provided by Rocky 9 (ipxe-roms, ipxe-bootimgs-*), among other things to support compressed images via imgextract. This is working well, only it seems like the download (HTTP) is no longer using the full 10G connection, instead it seems more like a 1G connection from an imprecise counting-in-my-head timing, so I'm suspecting that the interface is negotiated down to 1G. This was not the case with iPXE 1.0.0+

What would be good first steps in figuring out what is going wrong here? For reference, this is running `master` and booting on an x86 machine with a `BCM57414 NetXtreme-E 10Gb/25Gb RDMA Ethernet Controller`

EDIT: Another observation: It doesn't matter whether I use `initrd` or `imgextract`, both are slow. `ifstat` shows for both 1.0.0 and `master` build from source the same `net0: ... using NII on NII-... (open)` output

**Comment by stappersg** [comment]
> What would be good first steps in figuring out what is going wrong here?

Check on the network switch the port speed.

**Reply by upsj** [reply]
Here the relevant lines from `show interface`, with auto-negotiation off, I'm not sure if I would expect a difference/the switch to be able to recognize link speeds.
```
Pluggable media present, SFP28 type is SFP28 25GBASE-CR-2.0M
LineSpeed 10G, Auto-Negotiation off
Rate Info(interval 30 seconds):
  Output 338 Mbits/sec, 27442 packets/sec, 3% of line rate
```
I just double-checked, the image pull with 1.0.0 is indeed more than 5x faster, in line with what I would expect comparing a 1G compressed download vs. a 10G uncompressed download

**Comment by mcb30** [comment]
Try using https://ipxe.org/howto/bisect

**Reply by upsj** [reply]
that worked, after I figured out a working base commit, thanks!

**Comment by NiKiZe** [comment, marked as answer]
So both "working" and "non working" shows NII in iPXE ifstat ?

There is other discussions on buffer size changes causing slowdowns on BCM NICs, especially when used on links with some latency (wan)

**Reply by upsj** [reply]
The provisioned node and provisioning node are connected to the same switch, I get around 0.3ms ping latency, so I hope that that's not the issue 😄

**Reply by upsj** [reply]
You are onto something though, the offending commit is 2d180ce233f6eae8d0df279e5cbf134f547f604c "Update maximum window size to 2MB"

**Reply by NiKiZe** [reply]
So same/duplicate of #1023 ?

**Reply by upsj** [reply]
Yes, though it's less about being stuck and more about being slow from the start. By reverting https://github.com/ipxe/ipxe/commit/2d180ce233f6eae8d0df279e5cbf134f547f604c on master, I get the full speed I expect from the interface.

**Reply by mcb30** [reply]
> Yes, though it's less about being stuck and more about being slow from the start. By reverting [2d180ce](https://github.com/ipxe/ipxe/commit/2d180ce233f6eae8d0df279e5cbf134f547f604c) on master, I get the full speed I expect from the interface.

OK, that's very useful.  Could you try various other values for `TCP_MAX_WINDOW_SIZE` to see how your system behaves?

We can't just tweak `TCP_MAX_WINDOW_SIZE` without a justification, though.  We will need to understand and fix the root cause, which will most likely require looking at packet captures to understand what is happening at the TCP level.

**Reply by upsj** [reply]
I will likely not be able to spend much time on this soon, but I'll see if I can come back to it later
